src/lib/VirtualScroller.js

```diff
@@ -0,0 +1,183 @@
+        this.totalViewItems = Math.ceil(this.containerHeight / this.itemHeight);
+
+        // The height of the buffer before virtual scroll renders the next set
+        this.maxBufferHeight = this.totalViewItems * this.itemHeight;
+
+        // The max number of items to render at any one given time
+        this.maxRenderedItems = Math.ceil(this.totalViewItems * 3);
+
+        // Create the scrolling container element
+        this.containerEl = document.createElement('div');
+        this.containerEl.className = 'vs-container';
+
+        // Create the true height content container
+        this.contentEl = document.createElement('div');
+        this.contentEl.className = 'vs-content-container';
+        this.contentEl.style.height = `${this.totalItems * this.itemHeight + this.marginTop + this.marginBottom}px`;
+
+        this.containerEl.appendChild(this.contentEl);
+        this.container.appendChild(this.containerEl);
+
+        this.renderItems();
+
+        this.bindDOMListeners();
+    }
+
+    /**
+     * Binds DOM listeners
+     *
+     * @return {void}
+     */
+    bindDOMListeners() {
+        this.containerEl.addEventListener('scroll', this.onScrollHandler);
+    }
+
+    /**
+     * Unbinds DOM listeners
+     *
+     * @return {void}
+     */
+    unbindDOMListeners() {
+        if (this.containerEl) {
+            this.containerEl.removeEventListener('scroll', this.onScrollHandler);
+        }
+    }
+
+    /**
+     * Handler for 'scroll' event
+     *
+     * @param {Event} e - The scroll event
+     * @return {void}
+     */
+    onScrollHandler(e) {
+        const { scrollTop } = e.target;
+
+        if (Math.abs(scrollTop - this.previousScrollTop) > this.maxBufferHeight) {
+            // The first item to be re-rendered will be a totalViewItems height up from the
+            // item at the current location
+            const firstIndex = Math.floor(scrollTop / this.itemHeight) - this.totalViewItems;
+            this.renderItems(firstIndex < 0 ? 0 : firstIndex);
+
+            this.previousScrollTop = scrollTop;
+        }
+    }
+
+    /**
+     * Render a set of items, starting from the offset index
+     *
+     * @param {number} offset  - The offset to start rendering items
+     * @return {void}
+     */
+    renderItems(offset = 0) {
+        let count = this.maxRenderedItems;
+        // If the default count of items to render exceeds the totalItems count
+        // then just render the difference
+        if (count + offset > this.totalItems) {
+            count = this.totalItems - offset;
+        }
+
+        let numItemsRendered = 0;
+        const fragment = document.createDocumentFragment();
+
+        while (numItemsRendered < count) {
+            const rowEl = this.renderItem(offset + numItemsRendered, this.itemHeight);
+            fragment.appendChild(rowEl);
+            numItemsRendered += 1;
+        }
+
+        this.contentEl.innerHTML = '';
```

Comment:
I'm finding conflicting information on  the performance difference between clearing content via innerHtml vs. looping/removing DOM nodes. In most cases it wouldn't matter, but here it might make a difference if the latter is significantly faster in most browsers.

Take a look at https://stackoverflow.com/a/3955238